contracts/oracle/PCVOracle.sol

```diff
@@ -0,0 +1,226 @@
+    /// (implicit) disallow deposit being called if contract does not have PCV deposit role
+    function addIlliquidVenues(address[] calldata illiquidVenuesToAdd)
+        external
+        onlyGovernor
+    {
+        uint256 illiquidVenueLength = illiquidVenuesToAdd.length;
+        for (uint256 i = 0; i < illiquidVenueLength; ) {
+            _addIlliquidVenue(illiquidVenuesToAdd[i]);
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @notice add liquid venues to the oracle
+    /// only callable by the governor
+    /// 1. add venues
+    /// 2. governance action or later should call deposit, thus benchmarking the previous balance
+    /// (implicit) disallow deposit being called if contract does not have PCV deposit role
+    function addLiquidVenues(address[] calldata liquidVenuesToAdd)
+        external
+        onlyGovernor
+    {
+        uint256 liquidVenueLength = liquidVenuesToAdd.length;
+        for (uint256 i = 0; i < liquidVenueLength; ) {
+            _addLiquidVenue(liquidVenuesToAdd[i]);
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @notice set the market governance oracle address
+    /// only callable by governor
+    /// @param _marketGovernanceOracle new address of the market governance oracle
+    function setMarketGovernanceOracle(address _marketGovernanceOracle)
+        external
+        onlyGovernor
+    {
+        address oldMarketGovernanceOracle = marketGovernanceOracle;
+        marketGovernanceOracle = _marketGovernanceOracle;
+
+        emit MarketGovernanceOracleUpdated(
+            oldMarketGovernanceOracle,
+            _marketGovernanceOracle
+        );
+    }
+
+    /// ------------- Helper Methods -------------
+
+    function _afterActionHook() private {
+        MarketGovernanceOracle(marketGovernanceOracle).updateActualRate(
+            getLiquidVenuePercentage()
+        );
+    }
+
+    function _updateIlliquidBalance(int256 pcvDelta) private {
+        uint256 oldLiquidity = lastIlliquidBalance;
+
+        if (pcvDelta < 0) {
+            lastIlliquidBalance -= (pcvDelta * -1).toUint256();
+        } else {
+            lastIlliquidBalance += pcvDelta.toUint256();
+        }
+
+        emit IlliquidVenuePCVUpdated(oldLiquidity, lastIlliquidBalance);
+    }
+
+    function _updateLiquidBalance(int256 pcvDelta) private {
+        uint256 oldLiquidity = lastLiquidBalance;
+
+        if (pcvDelta < 0) {
+            lastLiquidBalance -= (pcvDelta * -1).toUint256();
+        } else {
+            lastLiquidBalance += pcvDelta.toUint256();
+        }
+
+        emit IlliquidVenuePCVUpdated(oldLiquidity, lastLiquidBalance);
+    }
+
+    function _addIlliquidVenue(address illiquidVenue) private {
+        illiquidVenues.add(illiquidVenue);
+
+        emit IlliquidVenueAdded(illiquidVenue);
+    }
+
+    function _addLiquidVenue(address liquidVenue) private {
+        liquidVenues.add(liquidVenue);
```

Comment:
that should be a separate action, in a function that allows us to take the sad path if the contract is older and will not be redeployed to conform to the happy path flow.

happy path for PCV Oracle:

1. PCV deposit contract gets created, deposit function reverts because calling into the PCV Oracle fails because the deposit does not have the `LIQUID_PCV_DEPOSIT` or `ILLIQUID_PCV_DEPOSIT` role. Thus PCV deposit has no PCV
2. PCV Deposit is granted either `LIQUID_PCV_DEPOSIT` or `ILLIQUID_PCV_DEPOSIT` role.
3. PCV Deposit is added in the PCV Oracle.
4. PCV is transferred to the deposit. Negative delta is recorded on the contract the withdraw happens from.
5. `deposit()` function is called on the PCV Deposit. Positive delta is recorded on the contract the deposit went into.

The benefit of making deposits in this manner is that once the `LIQUID_PCV_DEPOSIT` or `ILLIQUID_PCV_DEPOSIT` role is revoked, then PCV can no longer go into the underlying venue, and movements of PCV are automatically registered within the PCV deposit.